The four extensions in one package "extras"

bowman2001 opened the pull request:
Closes #12

jmooring commented:
The example from the README contains this:

```text
Hydrogen (H) is the 1^st^ element in the periodic table.

Water (H~2~O) is a liquid.

Water (H~2~O) is a ++transparent++ liquid.
```

The subscripts are failing.

This works:

```text
x H~2~O
```

This does not:

```text
x H~2~O x
```

bep commented on extras/subscript.go:
```diff
@@ -0,0 +1,115 @@
+}
+
+func (p *subscriptDelimiterProcessor) IsDelimiter(b byte) bool {
+	return b == '~'
```
This is a general comment that spans all insert/subscript/superscript.go files:

I have skimmed the implementation of all of these 3 and they look almost exactly the same, the only diff seems to be the delimiter and the HTML wrapper element.

I don't mind code duplication, but in this case I think we can make everything simpler by wrapping all of these into one extension/parser/renderer.

bep commented on extras/subscript.go:
```diff
@@ -0,0 +1,115 @@
+var defaultSubscriptParser = &subscriptParser{}
+
+// NewSubscriptParser returns a new InlineParser that parses subscript expressions.
+func NewSubscriptParser() parser.InlineParser {
```
This is related to the above (one parser), but I suggest that we for starters unexports everything that's not related to getting a extension.

bep commented on extras/subscript.go:
```diff
@@ -0,0 +1,115 @@
+}
+
+// Subscript is an extension that allows you to use a subscript expression like 'H~2~O'.
+var Subscript = &subscript{}
```
This works, but there will (maybe) eventually come some config, and I suspect that this becomes easier to document across packages if we have a conistent API. So I suggest we do something ala:

```go
extensions := extras.New(
    extras.SubscriptConfig{
            Enable: true
     },
     extras.SuperscriptConfig{
            Enable: true
     },
     ....
}
```

With my comment above about unexporting things, we could probably add something ala this later:

```go
extensions := extras.New(
    extras.SubscriptConfig{
            Enable: true
            HTMLRenderer: ...
     },
     ....
}
```

bowman2001 replied:
I'm using an enumeration type to identify and configure the extra inline tags with internal structs. I hope this is in the line of what you meant.

bowman2001 commented:
Yes, @bep, the only difference between Superscript and Subscript are the markers (a single '^' or '~'). All the other variables, functions, or methods only have different names. This redundancy could surly be removed. 

The insert doesn’t need to watch out for whitespace and has the same logic as strong emphasis. 

But first I need to address the problem @jmooring mentioned. My unit tests did not cover the cases "x H~2~O x" or "x foo^3^bar x" until now. They both fail, and I obviously missed something I need to figure out.

jmooring commented on extras/go.mod:
```diff
@@ -0,0 +1,5 @@
+module github.com/gohugoio/hugo-goldmark-extensions/extras
+
+go 1.22.1
```
```suggestion
go 1.20
```

bowman2001 commented:
Sorry for the delay, I had to attend to a lot of changes in my life in the last weeks. Coming month my new full-time Job is going to start. 

I want to finish the extras package with the requested changes in the next days.

bowman2001 commented:
There are new unit tests for the problem @jmooring found, and it's gone. The redundancy in the extensions is also gone. Only the necessary general structures for an inline tag remain which can be used for different tags. I have added a new mark tag because it was so easy: ==marked== -> <mark>marked</mark>. 

The configuration follows now the same concept as `passthrough` without any additional configuration data, as @bep requested.

I also changed `config.go` and `convert.go` in my Hugo fork and wrote some integration tests in `goldmark_integration_test.go`. Everything looks good.

I noticed that I am not as proficient with the Go workflow as I would like. I had to change the module path to my personal repo to integrate.

jmooring commented:
To simplify integration testing I am merging this without tagging a release. The repository is still described as "Work in progress".
